cachi2/core/models/sbom.py

```diff
@@ -445,6 +555,86 @@ def _unique_packages(cls, packages: list[SPDXPackage]) -> list[SPDXPackage]:
+        """Return non-root packages."""
+        return [p for p in self.packages if p.SPDXID != self.root_id]
+
+    @staticmethod
+    def retarget_and_prune_relationships(
+        from_sbom: "SPDXSbom",
+        to_sbom: "SPDXSbom",
+    ) -> list[SPDXRelation]:
+        """Retarget and prune relationships."""
+        out = []
+        for r in from_sbom.relationships:
+            # Do a copy to ensure we are not pulling a carpet from underneath us:
+            new_rel = r.model_copy(deep=True)
+            if new_rel.spdxElementId == from_sbom.root_id:
+                new_rel.spdxElementId = to_sbom.root_id
+            if new_rel.relatedSpdxElement == from_sbom.root_id:
+                new_rel.spdxElementId = to_sbom.SPDXID
+            # Old top-level "DESCRIBES" must go:
+            if not (
+                new_rel.relatedSpdxElement == from_sbom.root_id
+                and new_rel.relationshipType == "DESCRIBES"
+            ):
+                out.append(new_rel)
+        return out
+
+    def __add__(self, other: Union["SPDXSbom", Sbom]) -> "SPDXSbom":
+        if isinstance(other, self.__class__):
+            # Packages are not going to be modified so it is OK to just pass references around.
+            merged_packages = self.packages + other.non_root_packages
+            # Relationships, on the other hand, are amended, so new
+            # relationships will be constructed.
+            # Further, identical relationships should be dropped.
+            # Deduplication based on building a set is considered safe because all
+            # fields of all elements are used to compute a hash.
+            # Just using set won't work satisfactory since that would shuffle relationships.
+            merged_relationships, seen_relationships = [], set()
+            processed_other = self.retarget_and_prune_relationships(from_sbom=other, to_sbom=self)
+            for r in self.relationships + processed_other:
+                if r not in seen_relationships:
+                    seen_relationships.add(r)
+                    merged_relationships.append(r)
+            # The same as packages: files are not modified, so keeping them as is.
+            # Identical file entries should be skipped.
+            merged_files = list(set(self.files + other.files))
+            res = self.model_copy(
+                update={
+                    # At the moment of writing pydantic does not deem it necessary to
+                    # validate updated fields because we should just trust them [1].
+                    "packages": self.deduplicate_spdx_packages(merged_packages),
+                    "relationships": merged_relationships,
+                    "files": merged_files,
+                },
+                deep=True,
+            )
+            return res
+        elif isinstance(other, Sbom):
+            return self + other.to_spdx(doc_namespace="NOASSERTION")
+        raise Exception("Something smart goes here")
+
```

Comment:
I think there's already enough being done in this patch, so for better readability and for clarity purposes I'd leave the commit at just declaring the necessary models (unless you want to go down the road of even more granularity and introduce them gradually) and add this extra logic in a different patch, it's already very hard to follow without all the SPDX spec context.